Tower Screen enhancements

syntaxi opened the pull request:
Enhances the tower screen to include extra functionality and improve layout/visuals

* Adds formatting for the fields and values of a block
* Adds cores to the screen

Commit:
Rearrange `TowerInfoScreen` to allow for the addition of core information, block settings and general info

Commit:
Fix ui upgrade incorrectly calculating height and not centering text

Commit:
Fix changing label names and not updating them

Commit:
Implement cores into the tower screen

Commit:
Split UIComponentFields into three columns, field names, field values and upgrade values

Commit:
Add an interface to convert fields to more readable formats

Commit:
Change UpgradingSystem to use parsers when retriving values.
Also change to use `List<String>` not `Map<String, String>` for values.
This does require the lists to maintain order.

Commit:
Convert `UIUpgrader` to use the changes in `UIComponentFields`and `UpgraderSystem`

Commit:
Clear selected upgrade when different block selected.

Commit:
Change formatting in `UIComponentFields` to display centred and without wrapping.

Commit:
Javadoc for changes in prior commits.

Commit:
Add information on how much power a tower is using vs how much it is producing

Commit:
Refactor `TowerInfoScreen` to store the `TowerComponent` for a tower entity, instead of the `EntitiyRef` as it was only ever used to retrive the `TowerComponent` anyway

Commit:
Quick change to name of paramater in `BaseParser`

Commit:
Use order of field names in component values, not of the human readable alternatives

Commit:
Change thrown error in `UpgradingSytem#setField` from `Error` to `IllegalArgumentException`

Commit:
Add defaults for the handle methods in BaseParser

Commit:
Utilise methods in `java.lang.Invoke` to call methods on the parsers that are named the same as the field in question

Commit:
Javadoc for the prior commit's changes to `UpgradingSystem`

Commit:
Add in parsers for all the effector types.

Commit:
Fix dumb error where default values were handled with a flipped `isUpgrade`

Commit:
Make `BaseParser` an abstract class not an interface as it is providing number default implementations now

Commit:
Convert upgrading system to allow support of more types in upgrade and component fields.

Commit:
Add support for String's and Enums

Commit:
Give up on trying to cleverly handle types and simply convert them all to string and back again.

Commit:
Add `toString` methods to the `SelectionMethod` enum

oniatus commented on src/main/java/org/terasology/gooeyDefence/TowerManager.java:
```diff
@@ -87,6 +87,61 @@ private static long getTargeterId(String eventId) {
+     */
+    public static int getEffectorDrain(TowerComponent towerComponent) {
+        int drain = 0;
+        for (EntityRef effectorEntity : towerComponent.effector) {
```
would plural be better for the field name in towerComponent?

The three accumulator-methods can probably be written as a stream.map(.map).sum operation, however this is more a style than a worse/better question ;) The way it is is also okay.

syntaxi replied:
Now that you've pointed out the lack of a plural I can't unsee it. I'll fix it directly on the main branch after this is merged however, because it'll touch a lot of files.

I've also changed the methods to use map. After looking at it it looks cleaner to do it that way.

Commit:
So apparently eveything is either given as a Double or can be turned into a double.
Hence, I just convert all types into a double and then into the correct type (bar strings & enums)

Commit:
Add parsers for the targeters.

Commit:
Add in a string to indicate the type of block selected.

Commit:
Add method to set the targeting method

Commit:
Add buttons to select the targeting method

Commit:
Convert `getTargeterDrain` and similar core/effector methods to use `stream().map().sum()`
